doc/decisions/ensure.md

```diff
@@ -0,0 +1,68 @@
+# Ensure
+
+## Problem
+
+Applications want to ensure that some functionality (=global plugin)
+is present in Elektra.
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+- Keep `kdbEnsure` (Rejected because it is too flexible and can be called many times. Furthermore,
+  `kdbOpen` would build up configurations that get removed afterwards.)
+  - reduce for only global plugins, as the partial other functionality is confusing
+    and not needed
+  - find solution that list plugin is not needed
+  - only as implementation detail below libraries (e.g. like done for notification)
+- Specific APIs per plugin, Rejected because:
+  - difficult for application developers
+  - every plugin would need to design new APIs
+- Have a new API: `KDB * kdbOpen (Key * parent);`
+  `int kdbConfigure (KDB * handle, KeySet * contract, Key * parentKey);`
+  `KDB * kdbOpenDefault (Key * parent);`
+  The new kdbOpen only does the absolute minimum work, in particular it doesn't set up any global plugins.
+  If you use kdbOpen you must call kdbConfigure otherwise kdbGet will fail. kdbConfigure configures global
+  plugins (basically just a renamed kdbEnsure). Lastly, kdbOpenDefault does more or less what the old kdbOpen
+  does. It sets up the default case and you can call kdbGet immediately. But you cannot call kdbConfigure after
+  kdbOpenDefault.
+  Rejected because of API bloat and introduction of further state in `kdb`.
+
+## Decision
+
+Integrate `kdbEnsure` in `kdbOpen(Key *errorKey, KeySet *contract)` but only allow global plugins.
+
+## Rationale
+
+- can immediately build up correct plugin positioning
+- does not allow to start applications if the contract cannot be fulfilled
+- simplest and minimalistic solution
+
+## Implications
+
+`elektraNotificationOpen` needs to be removed and instead we need a new API:
+
+```c
+KeySet * contract = ksNew (0, KS_END);
+elektraNotificationGetContract (contract);
+KDB * kdb = kdbOpen (key, contract); // contract contains notification config
+elektraIoSetBinding (kdb, binding);
+```
+
+Similar contract getters need to be introduced for other functionality, like gopts:
+
+```c
+KeySet * contract = ksNew (0, KS_END);
+elektraGOptsGetContract (contract, argc, argv, environ);
+KDB * kdb = kdbOpen (key, contract);
+```
+
+The high-level API can make this more pretty, though.
```

Comment:
Not 100% sure this is the best possible API, but yes something like this is what I thought of.

The proposal is a bit inconsistent. AFAICT `elektraIoSetBinding` is a separate function to avoid having a `Key` whose value is a `ElektraIoInterface *`. But then there is `elektraGOptsGetContract` which needs to store `char **`, which has the same problems.

So either, both cases use a separate setter function called after `kdbOpen` that directly manipulates the relevant `Plugin` struct, or they both just use a `*GetContract` function that creates binary keys.

Personally, I think a separate getter makes more sense. The whole point of having a `contract` KeySet is that the API/ABI doesn't have to change if additional config options are added. A `elektraNotificationGetContract` function would have to change, if e.g. notifications get some new config option.

Reply:
> Not 100% sure this is the best possible API,

Also not yet sure, here is an improvement: 581c6bf3018bea540343de30f342e948c781d772

Reply:
Yes, the new API looks a lot better. However, I would prefer, if either `kdbOpen` took ownership of the `KeySet * contract` or at least copied what is needed, so that `ksDel (contract)` can be called right after `kdbOpen`. Otherwise there is too much potential for memory leaks IMO.

Reply:
Yes, of course it will copy, so ksDel afterwards is safe: a6429dc9a89ba25340f3e72b2883daff85b00333

Taking ownership can easily lead to misunderstandings, especially because KeySet is different to Key (which is ref-counted, so the ownership can be shared even if "taken" like in ksNew). In elektraPluginOpen we have such an API and I actually think we should change it there (if we decide to make this API public).